File: examples/tictactoe/web/src/stores/BoardStore.ts
import { TriggerStore } from '@boatkit-io/restream';
import { AutoSubscribeStore, autoSubscribe } from '@boatkit-io/resub';

import { BoardStoreName, BoardStoreState, BoardStoreStatePartial } from '../restream/PackageMain';

@AutoSubscribeStore
class BoardStore extends TriggerStore<BoardStoreState> {
    constructor() {
        super(BoardStoreName, BoardStoreState, BoardStoreStatePartial);
    }

    @autoSubscribe
    getBoard(): string[][] {
        return (this._state.board ?? []).map((row) => row ?? []);
    }

    @autoSubscribe
    getXTurn(): boolean {
        return this._state.xTurn;
    }
}

export default new BoardStore();

File: examples/tictactoe/internal/game/boardstore.go
package game

import (
	"errors"
	"reflect"

	"github.com/boatkit-io/restream/pkg/restream"
)

// @restream.store(BoardStore)
type BoardStore struct {
	storeData *restream.StoreData[BoardStoreState, *BoardStoreState, *BoardStoreStatePartial]
}

func NewBoardStore(rpcd *restream.RPCDispatcher) (*BoardStore, error) {
	s := &BoardStore{}

	initialState := &BoardStoreState{
		Board: [][]string{
			{"", "", "O"},
			{"", "X", ""},
			{"O", "", "X"},
		},
		XTurn: true,
	}

	s.storeData = restream.NewStoreData[BoardStoreState, *BoardStoreState, *BoardStoreStatePartial](s, initialState)

	rpcd.RegisterRPCHandler("PlaceToken", 1, func(x, y int) error {
		partial := &BoardStoreStatePartial{
			Board: restream.NewPartialArray[[]string](),
		}
		state := s.storeData.ReadState()
		newRow := append([]string{}, state.Board[y]...)
		xTurn := state.XTurn
		if newRow[x] != "" {
			return errors.New("cell already occupied")
		}
		if xTurn {
			newRow[x] = "X"
		} else {
			newRow[x] = "O"
		}
		partial.Board.Set(y, newRow)
		partial.XTurn = restream.Ptr(!xTurn)
		s.storeData.ApplyPartial(partial)
		return nil
	}, reflect.TypeFor[PlaceTokenRequest](), reflect.TypeFor[PlaceTokenResponse]())

	return s, nil
}
